pyuvdata/uvdata/ms.py

```diff
@@ -1165,6 +1166,40 @@ def write_ms(
 
             ms.putcol("FIELD_ID", np.repeat(field_id_array, self.Nspws))
 
+            # If scan numbers are not already defined from reading an MS,
+            # group integrations (rows) into scan numbers.
+            if self.scan_number_array is None:
+
+                slice_list = []
+                for idx in range(self.Nphase):
+                    sou_id = self.phase_center_catalog[sou_list[idx]]["cat_id"]
+                    slice_list.extend(
```

Comment:
Took me a second to parse what this piece of code was doing, but this is spiffy! I actually wonder if this should actually be a function with the `utils` module (or maybe even the `UVData` object? Would be curious what @bhazelton might say here), since I could see this code being used outside of this specific module (e.g., I might similarly want to use in in `mir`).

Reply:
If it's useful outside of this context we should definitely pull it out. Whether as a utility function or a private method on UVData depends on whether it's likely to be useful for other objects and how many other object attributes contribute to it. I think you're best placed to decide between those @kartographer.

Reply:
In that case, I think I might vote to make it a private method under `UVData` -- even if we add a scan number attribute to `UVFlag`, `UVCal`, etc., I expect those values to be imported from a `UVData` object rather than being derived in the way they are here.

Reply:
Thanks for the input! I've moved this code to `UVData._set_scan_numbers`.